### PR TITLE
Custom Dockerfile for gitpod.io

### PR DESCRIPTION
**Pull request by Nuxij**
Little configuration to give https://gitpod.io users an environment with the SDKs pre-installed

Gitpod is an amazing tool, even without this PR being accepted, it's worth your time to have a peek!

### CODE REVIEW

**Comment by pfeerick** [comment]
Lol... you beat me to it... that was going to be my next step after the changes from #137 went through ;)

**Comment by JF002** [comment]
I don't know gitpod. What is it? An IDE in the cloud? A CI?

**Comment by Nuxij** [comment]
It's a hosted IDE but each instance of that ide, a "workspace", has access to a whole "host system" because it's just a docker container.

You can use it for free on any public repo without these changes, just by appending your URL. e.g. https://gitpod.io/#https://github.com/JF002/Pinetime

This change commits a Dockerfile (and config to tell it) so that "workspaces" run in our custom crafted environment (i.e. has the SDKs like the normal docker container would)

Once that's set up it does have a concept of pre-building said workspace images at various times (like for PRs or branches etc), so it is a CI solution in effect too, yes. I'm not bothered by that bit right now though as we use Github Actions right now anyway. It's just a cool side feature! 

Open Source projects get 50 hours free build time on the platform. In my testing this appears to be actual CPU usage "time". Hours of using the editor does nothing to the clock, but a few compiles and my clock is now at 48.5 or something.

**Comment by pfeerick** [comment]
As Joe said, it's basically an online IDE and workspace - using that gitpod.io URL, anyone could edit and build the code on any internet connected device - nothing needs the be installed locally to be able to compile. Another example would be the [ISS-Notifier](https://github.com/pfeerick/ISS-Notifier) project I migrated from Arduino IDE -> PlatformIO and am slowly adding more features to... you can use the badge at the top of the readme to [spin up a gitpod instance](https://gitpod.io/#https://github.com/pfeerick/ISS-Notifier) - which automatically prepares the build environment. So I (or anyone else) can make changes, type `pio run` and it'll rebuild. Easy peasy! :)

**Comment by JF002** [comment]
From what I can see in my account, I "only" have 50hours.

`It's a hosted IDE but each instance of that ide, a "workspace", has access to a whole "host system" because it's just a docker container.`
Does it mean that the docker container runs on my computer? Is it possible to flash and debug the target (pinetime connected on a SWD debugger) directly from Gitpod?

Or do we just edit and build the code online, fetch the binary to try them on the actual PineTime locally?

In order to understand, I tried to run your branch : https://gitpod.io/#https://github.com/Nuxij/Pinetime/tree/gitpod-config but it stops with an error after a few seconds : 

```
building base image

initializing base image context
UGit operations complete                       location=/workspace/init
running base image build
Step 1/11 : FROM gitpod/workspace-full
latest: Pulling from gitpod/workspace-full
Digest: sha256:30b68211fc43aae9629d7c4e82f5ab37c296dc0b2317d0ee64b89943fa4038a9
Status: Image is up to date for gitpod/workspace-full:latest
 ---> 255ce48e62b9
Step 2/11 : ARG DEBIAN_FRONTEND=noninteractive
 ---> Using cache
 ---> 44e5f3acae50
Step 3/11 : RUN apt-get update -qq     && apt-get install -y       bash       build-essential       cmake       git       make       python3       python3-pip       tar       unzipwget       libffi-dev       libssl-dev       python3-dev     && rm -rf /var/cache/apt/* /var/lib/apt/lists/*;
 ---> Running in 4b7cb28fe613
E: Could not open lock file /var/lib/apt/lists/lock - open (13: Permission denied)
E: Unable to lock directory /var/lib/apt/lists/



build failed: cannot build base image: The command '/bin/sh -c apt-get update -qq     && apt-get install -y       bash       build-essential       cmake       git       make       python3       python3-pip       tar       unzip       wget       libffi-dev       libssl-dev       python3-dev     && rm -rf /var/cache/apt/* /var/lib/apt/lists/*;' returned a non-zero code: 100

Error: build failed: cannot build base image: The command '/bin/sh -c apt-get update -qq     && apt-get install -y       bash       build-essential       cmake       git       makepython3       python3-pip       tar       unzip       wget       libffi-dev       libssl-dev       python3-dev     && rm -rf /var/cache/apt/* /var/lib/apt/lists/*;' returned a non-zero code: 100
```

**Comment by pfeerick** [comment]
No, it runs on their end (the 'hosted' bit). I doubt debugging will be
practical /possible because of that. Basically means anyone can build the
files without having to even install anything.

The 50 vs 100 hours is curious... Mine says 100, and I'm not using a plan
or anything either.

**Comment by pfeerick** [comment]
Oh, I could be talking complete nonsense, but I seem to think something
just changed with gitpod and permissions, docker, etc... But that could
also have been three months ago, whatever... Was reading their news posts
the other week and don't remember the 'when' off hand...

**Comment by Nuxij** [comment]
As far as I know, it runs on their servers. You can run your own instances locally if you like though.
As for flashing, it won't have any direct local access as it's just a web browser cloud ide, but they do have support for exposing http ports from your workspace so we could utilise that for download/remote flash url.

**Comment by pfeerick** [comment]
The errors will be because Gitpod effectively runs the dockerfile as USER Gitpod, not root... so sudo needs to be added to the apt update, install and rm on line 4, and probably the pip install on line 24. The toolchain installs will probably be fine though?

You can also find the compiled firmware binary in the code editor file browser, and right click to download it ;)

image

It looks like the pricing table on the gitpod website ... could be laid out a bit better ... it's 50 hours per month, or 100 hours if you do open source development... which I must have automatically met the criteria for as I don't remember applying for it. 

image

File: docker/.gitpod.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,38 @@
+FROM gitpod/workspace-full
+
+ARG DEBIAN_FRONTEND=noninteractive
+RUN apt-get update -qq \
```
Comment (pfeerick) [inline docker/.gitpod.Dockerfile:4]: Adding  `sudo` to `apt-get update`, `apt-get install`, and `rm` gets past the initial error.

File: docker/.gitpod.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      unzip \
+      wget \ 
+# aarch64 packages
+      libffi-dev \
+      libssl-dev \
+      python3-dev \ 
+    && rm -rf /var/cache/apt/* /var/lib/apt/lists/*;
+
+# Needs to be installed as root
+RUN pip3 install adafruit-nrfutil
```
Comment (pfeerick) [inline docker/.gitpod.Dockerfile:24]: If this needs to be installed as root, probably needs `sudo` also

File: docker/.gitpod.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+RUN sudo chown -R gitpod /opt
+
+COPY docker/build.sh /opt/
+# Lets get each in a separate docker layer for better downloads
+# GCC
+RUN bash -c "source /opt/build.sh; GetGcc;"
+# NrfSdk
+RUN bash -c "source /opt/build.sh; GetNrfSdk;"
+# McuBoot
+RUN bash -c "source /opt/build.sh; GetMcuBoot;"
```
Comment (pfeerick) [inline docker/.gitpod.Dockerfile:35]: This uses pip3 behind the scenes, but didn't error for me, so maybe this one is fine?

**Review by pfeerick — COMMENTED** [review]
I tried a variation of this, https://github.com/pfeerick/InfiniTime/tree/gitpod, and it at least built without error. My question then would be (and to some extent, still is, as I'm not convinced all the bugs... er... kinks are ironed out yet)... how do I use it? What commands do I run to actually build now, since it's not being done automatically via the `build.sh` like normally done with docker.

**Comment by JF002** [comment]
@pfeerick variation loaded successfully in my gitpod (https://gitpod.io/# https://github.com/pfeerick/InfiniTime/tree/gitpod).

I'm not sure this is how it is intended to by used, but I could build the project by typing the following commands in the console : 

mkdir build
cd build
cmake -DARM_NONE_EABI_TOOLCHAIN_PATH=/opt/gcc-arm-none-eabi-9-2020-q2-update -DNRF5_SDK_PATH=/opt/nRF5_SDK_15.3.0_59ac345 -DUSE_OPENOCD=1 ../
make -j

Files generated by the build are available in the source tree (build/src) as expected.

image

**Comment by Nuxij** [comment]
Alright lets try a rebase to reduce the change set a bit!

File: docker/.gitpod.Dockerfile (hunk 1 above)
Comment (Nuxij) [reply docker/.gitpod.Dockerfile:5]: We can just bump the USER and drop it at the end

File: docker/.gitpod.Dockerfile (hunk 3 above)
Comment (Nuxij) [reply docker/.gitpod.Dockerfile:34]: Yeah I also see it not error, but I've wrapped all this bit as `USER root` anyway.
The `gitpod` user only really needs write access here if they want to run the Get... stuff during development (low likelihood?)

**Comment by Nuxij** [comment]
Ok, all sorted. Workflow is to load the workspace and run `/opt/build.sh`.

**Comment by pfeerick** [comment]
:egg-on-face: Of course it is... duh! And thanks JF for the test and also working method of compiling... I ended up doing that a bit recently playing with the joaq fork...

**Comment by pfeerick** [comment]
Oh, and I can confirm spinning up https://gitpod.io/#https://github.com/Nuxij/Pinetime and doing a `/opt/build.sh` *just worked*:tm:, and I could then quite easily have downloaded the binaries ready to flash. 

I avoided mentioning the use of user root as I saw this in the [gitpod documentation](https://www.gitpod.io/docs/config-docker/)
> When you are launching the Gitpod IDE, the local console will use the gitpod user, so all local settings, config file, etc. should apply to /home/gitpod or be run using USER gitpod (we no longer recommend using USER root).
>
> You can however use sudo in your Dockerfile. 

But for this it probably doesn't really matter.

**Comment by Nuxij** [comment]
Agreed, thanks for getting stuck in @JF002 🙂 

On the `USER root` idea I'm with ya. Both ways are essentially the same so whatevs 👍

**Comment by JF002** [comment]
@pfeerick you probably meant https://gitpod.io/#https://github.com/Nuxij/Pinetime/tree/gitpod-config ?

So, yeah, it works :) Thanks for your patience during this long review :)

**Comment by pfeerick** [comment]
Yeah, that will be it... I was on the main branch page when I copied that in... don't usually need the URL since I use an extension that adds a gitpod button to repos... Glad to see it's in now.

**Pull request merged**